OneSignalSDK/onesignal/src/main/java/com/onesignal/OSSessionManager.java

```diff
@@ -209,39 +212,77 @@ void setLastNotificationsId() {
             }
         }
 
-        notificationIds = notificationsIds;
+        indirectNotificationIds = notificationsIds;
     }
 
     SessionResult getSessionResult() {
-        if (session == OSSessionManager.Session.DIRECT && notificationId != null) {
+        if (session.equals(Session.DIRECT) && isDirectSession()) {
+            // Make sure direct flag is true
             if (OutcomesUtils.isDirectSessionEnabled()) {
-                JSONArray notificationIds = new JSONArray().put(notificationId);
-
+                JSONArray directNotificationIds = new JSONArray().put(directNotificationId);
                 return SessionResult.Builder.newInstance()
-                        .setNotificationIds(notificationIds)
+                        .setNotificationIds(directNotificationIds)
                         .setSession(Session.DIRECT)
                         .build();
             }
-        } else if (session == OSSessionManager.Session.INDIRECT && notificationIds != null) {
+        } else if (session.equals(Session.INDIRECT) && isIndirectSession()) {
+            // Make sure indirect flag is true
             if (OutcomesUtils.isIndirectSessionEnabled()) {
                 return SessionResult.Builder.newInstance()
-                        .setNotificationIds(notificationIds)
+                        .setNotificationIds(indirectNotificationIds)
                         .setSession(Session.INDIRECT)
                         .build();
             }
         } else if (OutcomesUtils.isUnattributedSessionEnabled()) {
+            // Make sure unattributed flag is true
             return SessionResult.Builder.newInstance()
                     .setSession(Session.UNATTRIBUTED)
                     .build();
         }
 
+        // Outcomes is disabled if it gets here
         return SessionResult.Builder.newInstance()
                 .setSession(Session.DISABLED)
                 .build();
     }
 
-    private boolean directSessionRecentlySet() {
-        long directSessionResetTime = directSessionTime != null ? System.currentTimeMillis() - directSessionTime : MAX_DIRECT_SESSION_TIME_SET;
-        return directSessionResetTime < MAX_DIRECT_SESSION_TIME_SET;
+    /**
+     * Validate if the current session could be indirect
+     * If indirectNotificationIds exist and have at least 1 notification id, this would imply a
+     * app open with notifications within the attribution window time
+     */
+    boolean isIndirectSession() {
+        return getIndirectNotificationIds() != null
+                && getIndirectNotificationIds().length() > 0
+                && OneSignal.appEntryState.equals(OneSignal.AppEntryAction.APP_OPEN);
+    }
+
+    /**
+     * Validate if the current session could be direct
+     * With a directSessionTime and a directNotificationId set this would imply that the notification opened logic occurred
+     */
+    boolean isDirectSession() {
```

Comment:
Thinking this method should be called `isAppEntryFromNotificationClick` or something that includes entry instead of session.

Reply:
Also this method should only be used for `restartSessionIfNeeded` and `attemptSessionOverride ` as we only care about the entry action if it was a result of a new session OR overriding on entry.

Keep `addSessionNotificationsIds` and `getSessionResult` on the pre-existing logic, we want to report the current session and don't want to check how we last got into the app.

This is so if we click on a notification that opens the app, then background the app, then tap on the icon to resume it the session is still DIRECT and don't want the last `appEntryState` to effect it.

Reply:
Got it, fixed now